### PR TITLE
Debugger: Overhaul symbol table parsing (add .mdebug and .sndata support) and add inspector GUI for complex data types

### PR DESCRIPTION
**Pull request by chaoticgd**
### Description of Changes

I have imported symbol table parsing, symbol database and C++ AST code from [CCC](https://github.com/chaoticgd/ccc), entirely replacing the existing symbol map class, and implemented a GUI to inspect data structures in memory using this information.

The original commit history is on the [data_inspector_backup](https://github.com/chaoticgd/pcsx2/tree/data_inspector_backup) branch.

New screenshot:

Screenshot_20240314_184024

Old screenshot:

Screenshot_20231104_003704

Some highlights:
- CCC library rewritten to support importing symbol from multiple different symbols tables at once (e.g. some translation units will lack debugging symbols but will have linker symbols) and user-defined symbols.
  - The new symbols database uses a handle system for accessing symbols. This is useful for retaining persistent references to symbols.
  - Support for MIPS debug (.mdebug), SNDLL (.sndata) and ELF (.symtab) symbol tables. For information about these formats, see the readme linked above.
  - This takes up quite a bit more memory (on the order of 100MB) than what was there before when a full .mdebug symbol table is present. If it's too much it could be possible to only load the symbols when the debugger is opened, but I haven't implemented that so far.
  - ~~Do comment on whether it would be better for me to put this in pcsx2/DebugTools/ccc/ like I have now with the PCSX2 license headers, or in 3rdparty/ccc/ with my own license headers. The original library is MIT so the latter might make it easier for me to copy changes back and forth? It probably doesn't matter too much, so whatever is convenient is probably fine.~~ I've moved it.
- New SymbolGuardian class that controls access to the symbol database.
  - The SymbolMap class has been entirely removed and replaced with this.
  - When an ELF file is loaded a worker thread is created to parse the symbol tables. Parsing a fully loaded .mdebug symbol table usually takes about a second (on my PC anyway).
    - One odd side effect of this is if new code is loaded in the meantime (e.g. loading from a savestate), the result of analysis may depend on how long parsing the symbol table takes. I suspect a good solution to this would be let the user scan for functions on demand, but that seems slightly out of scope for this PR. I could make ScanForFunctions use the code from the ELF instead. Please comment on what would be preferable.
    - Some built-in types are automatically added to the database so that even without debug symbols the symbol trees are somewhat useful.
 - The symbol trees in the debugger GUI can be used to inspect complex data types in memory.
   - Uses Qt model/view system.
   - Options for adding/removing/modifying symbols using dialogs and delegates.
   - Options for grouping symbols by module/ELF section/source file.
   - Function hashing system to detect symbols that are no longer valid and gray them out.
- Added the GNU demangler to replace the Avast one that was there before.
  - This is required to support the GCC 2.x mangling scheme used in lots and lots of games.
  - Based on code from GCC 13.2.0 with the old GCC 2.x demangler (circa 2018) ported back into it.
  - Includes my own fix for a memory safety issue with the cplus_demangle_opname function.
  - A readme is included that describes how I came up with the code that I did.
- The disassembly manager doesn't really work very well with the new symbol database. For now I've patched it up so it at least mostly works.

Progress:
- [x] Refactor [CCC](https://github.com/chaoticgd/ccc) to have acceptable error handling for integration with PCSX2.
- [x] Import symbol table parsing and AST code from [CCC](https://github.com/chaoticgd/ccc).
- [x] Create a data inspector window in the Qt GUI.
- [x] Create a [QAbstractItemModel](https://doc.qt.io/qt-6/qabstractitemmodel.html) to inspect data structures in memory using this type information.
- [x] Implement the globals tab in the data inspector window.
- [x] Implement the stack tab in the data inspector window.
- [x] Rewrite CCC so it can replace the existing SymbolMap class.
- [x] Implement support for other kinds of symbol tables (.symtab, SNDLL) in CCC.
- [x] Merge the data inspector window into the debugger window.
- [x] Replace the SymbolMap class with the SymbolDatabase class from CCC.
- [x] Various improvements (better editor widgets, VS project files, etc).
- [x] Rebase everything.

Possible future extensions:
- Support for DWARF 1 symbol tables, as were written out by the Metrowerks compiler. This would allow for many more games to be supported by the data inspector. I think the current architecture I have with the C++ AST would allow for this (with some changes), but it would still probably be a long-term goal.
- Watch window. This was originally going to be part of this PR, but I've decided to skip it for now. Ideally it would have an expression parser for a C-like language.
- JSON import/export. This would allow people to save their user-defined symbols, and to transfer certain symbols (e.g. data types) between games.
- Undo/redo support.
- Parser for a subset of C++. This would allow for new data types to be created by the user. Something like [this](https://github.com/chaoticgd/wrench/tree/master/src/cppparser) would probably do.
- Import symbol tables from IRX modules and SNDLL files. Hook the loader functions, import/delete the symbols on the fly.

### Rationale behind Changes

Many games shipped with debug symbols in this format, so this would be an extremely useful addition for the modding communities of said games.

### Suggested Testing Steps

Some games with .mdebug symbols (data types, functions, globals):
- Alex Ferguson's Player Manager 2001
- European Tennis Pro
- Fatal Frame
- Go Go Golf
- Hard Hitter Tennis
- Jet X2O
- MTV Music Generator 2
- Orange Pocket: Root
- Sega Soccer Slam
- The Sims
- The Weakest Link

A game with SNDLL symbols (functions, globals):
- Ratchet & Clank: Size Matters

### CODE REVIEW

File: pcsx2-qt/Debugger/DataInspectorWindow.cpp
Comment (stenzek) [inline pcsx2-qt/Debugger/DataInspectorWindow.cpp:15]: Comment's correct, this is a race. This whole bit is - CDVD functions cannot be called anywhere other than the CPU thread.

Probably should do it on the CPU thread at ELF load time, like the current symbol stuff.

Comment (chaoticgd) [reply pcsx2-qt/Debugger/DataInspectorWindow.cpp:15]: I'm not sure what the best thing to for that is currently. The basic problem is that the AST in memory can be quite large (e.g. about 250MB in memory for The Sims, there are probably worse examples), so I didn't want it to be loaded if the data inspector wasn't opened. I could either load the symbols all the time, keep the unparsed symbol table in memory all the time (which is much smaller), or rework the ELF code so I can do what I'm doing now, but actually thread safe. In the former case I could see what I could do to make the AST smaller in memory. There are some trivial things I can think of, and some less trivial things.

Comment (stenzek) [reply pcsx2-qt/Debugger/DataInspectorWindow.cpp:15]: Then it should kick a request off to the CPU thread asynchronously, and pass it back to the UI thread. You can use `Host::RunOnCPUThread()` and `QtHost::RunOnUIThread()` to do so without creating additional entry points, just have to be careful if the window gets closed before the UI callback fires.

Comment (chaoticgd) [reply pcsx2-qt/Debugger/DataInspectorWindow.cpp:81]: Check the new version of the code. I've put the CDVD code on the CPU thread and spun off the symbol table parsing onto its own thread using QtConcurrent::run. Also, I went in with massif and got CCC's memory usage down by over 50%.

**Comment by refractionpcsx2** [comment]
I don't know if anybody else agrees, but is there any reason this couldn't be in the debugger? I feel like they are kind of doing similar jobs, and it would be nice to be able to right click on a variable and do "Go to in memory view" etc, all these features just feel like they would be at home as part of the debugger, rather than a separate screen.

Great features though!

**Comment by chaoticgd** [comment]
> I don't know if anybody else agrees, but is there any reason this couldn't be in the debugger?

There isn't any particular reason, especially now memory consumption is down a lot. That said, I wanted to focus on getting the functionality working first, then worry about integrating it with the existing debugger window. If anyone has more specific design ideas, certainly feel free to suggest them.

One constraint I have is that I think the watch window should at least be accessible alongside the rest of the debugger, so people can observe how values change as they step through code, just like with a regular GUI debugger. I think having it as a tab at the bottom would make sense, maybe with the option to pop it out as its own window.

Perhaps [Qt Dock Widgets](https://doc.qt.io/qt-5.15/qtwidgets-mainwindows-dockwidgets-example.html) or [KDDockWidgets](https://github.com/KDAB/KDDockWidgets) would make sense here?

**Comment by refractionpcsx2** [comment]
Yeah no worries.

I believe Fobes was thinking about changing the debugger so it was separate (docked?) windows, because the debugger is a bit cramped, so it would be nice to split it out, but I want expecting you to go that far with scope lol

**Comment by chaoticgd** [comment]
Got the stack tab working, with some caveats.

Screenshot_20231109_091444

The biggest issue is that for some functions the live ranges seem to be invalid.

It should be possible to display the values of parameters as well as locals, although the registers stored in the symbol table seem to be the registers used in the body of the function, not the prologue, so maybe there's something to be done there to make that more user friendly if you just break on the first instruction of a function.

**Comment by chaoticgd** [comment]
Some further thoughts about integrating this with the debugger window:

One challenge I see is how we're going to approach the multiple sources of truth that are the two different symbol tables. I see two different approaches to deal with this:
- Implement different tree models for the different symbol tables.
- Implement support for the standard ELF symbol table in CCC, and have it process both symbol tables into the same data structure in memory.

The latter would probably be cleaner, although also more invasive.

**Comment by refractionpcsx2** [comment]
I agree the latter sounds like a better option, try to keep the symbols unique, we don't *really* want multiple sources of truth if we can help it, but this is going to be true if it's integrated with the debugger or not.

**Comment by chaoticgd** [comment]
Little update: I'm still working on this. Building the new data structures and porting my existing code to use them is taking some time. I'll copy it all into this PR when it's all relatively stable.

File: 3rdparty/rcheevos/rcheevos
Comment (chaoticgd) [inline 3rdparty/rcheevos/rcheevos:1]: Whoops.

**Comment by chaoticgd** [comment]
This is finally ready for review and testing. Take as long as you want, I'm aware it's ballooned in scope quite a lot.

**Comment by Daniel-McCarthy** [comment]
Couple things I've found while giving it a more thorough test:
- It's crashing for me if I try to delete a local or parameter.
- Creating a parameter and local should be blocked if the cpu is dead, as typing an address in the dialog to create one during that state will cause a crash.
- `Right click -> Go to in Memory View` isn't quite working as expected. It doesn't set the location in memory on my end (the function in `MemoryView` isn't getting called when the signal is supposed to get called.\
\
If you right click a row on the Name or Value column and do `Go To in Memory View` it doesn't change the position in memory. It will if you right click on the Location column, but not actually because it only works for that column, but that if you left click or right click the location column it will instantly move the memory position to it correctly. 

Just wanted to give a heads up on these.

**Comment by weirdbeardgame** [comment]
The earlier issue I was having with Fatal Frame 1 is resolved.

**Comment by weirdbeardgame** [comment]
image
Okami it looks like the symbols table died

**Comment by Risae** [comment]
Fixes https://github.com/PCSX2/pcsx2/issues/10692 (Function column labels) 
Partially fixes https://github.com/PCSX2/pcsx2/issues/10678 (Rename Function)

**Review by Daniel-McCarthy — APPROVED** [review]
All issues I ran into are resolved. LGTM, haven't found any new issues. 👍
Great work on this

**Comment by chaoticgd** [comment]
I have actually found another issue, a recent regression with enum editors. Also wondering about what the best way to get Codacy to stop complaining is.

**Comment by chaoticgd** [comment]
That should be all the issues found so far sorted. I want all the people who regularly work on the debugger to have a look through the symbol guardian/symbol database stuff to make sure it's to their satisfaction.

As for the Okami thing, it looks like the game's symbol table is obfuscated so there's nothing to be done.

**Comment by chaoticgd** [comment]
Aand of course there were more bugs, which are now fixed. I'll continue playing about with it to see I can find any more.

**Comment by chaoticgd** [comment]
I've been thinking, and I've come to the conclusion that the current design for the SymbolGuardian class is unnecessarily complicated. Specifically how the worker thread that loads the symbol table(s) takes a lock on the symbol database for that entire duration, which leads to there being TryRead/BlockingRead/TryReadWrite/BlockingReadWrite functions, which are probably not all required.

A better design would be to build the symbol table in a separate SymbolDatabase object and then merge them at the end. I think I strayed away from this originally because I thought it would cause problems with conflicting symbol handles, but I've come up with a solution to that: Make the ccc::SymbolList<>::m_next_handle member variables `static std::atomic<u32>`. I'm not sure why, but it took me until today to think of this.

So far all the changes made since I said the PR was ready for review were fairly minor, although if this PR is going to take a long time to review anyway I think I may as well make the improvements immediately.

I haven't thought through how exactly all the aspects of this will work, I just want to get this down for now.

File: pcsx2-qt/Debugger/SymbolTree/SymbolTreeWidgets.cpp
Comment (Daniel-McCarthy) [inline pcsx2-qt/Debugger/SymbolTree/SymbolTreeWidgets.cpp:368]: It might be better UX to check if the user has a selection and conditionally add context menu options if they do. Currently it gives the option to rename/copy/goto even if no items exist.  [Here's an example](https://github.com/PCSX2/pcsx2/blob/master/pcsx2-qt/Debugger/CpuWidget.cpp#L325) where we do this.
 
By no means a big issue, just wanted to mention this.

**Comment by Daniel-McCarthy** [comment]
> So far all the changes made since I said the PR was ready for review were fairly minor, although if this PR is going to take a long time to review anyway I think I may as well make the improvements immediately.

I'm quite happy with the functionality in this PR, and the code is quite clean and easy to understand. From my side in it's current state I think it's in a good state to be merged, but I don't  have much say in that. Would be great to have @F0bes review it as well.

If you'd like to make those changes, feel free to. That said, it can always go in after this is merged as well.

**Review by Daniel-McCarthy — APPROVED** [review]
I've looked through all the code for this MR and am quite happy with it. The functionality has been refined with additional commits and I've been testing it. Looks great to me.

Haven't found any new issues. 👍

File: pcsx2-qt/Debugger/SymbolTree/SymbolTreeWidgets.cpp
Comment (chaoticgd) [reply pcsx2-qt/Debugger/SymbolTree/SymbolTreeWidgets.cpp:368]: I prefer disabling the menu items instead of removing them entirely so you can get a full picture of all the options that are available. That said, I don't do this for the goto in memory/goto in disassembly options currently so that's something I could add.

**Review by F0bes — APPROVED** [review]
I don't have any games with supported symbols, but a block-dump allowed me to experiment a little bit.
Overall everything looked good, and it wasn't lacking on the performance side.
You could've put ccc in the 3rdparty directory if you expect the need to keep ccc updated. But that's not a big deal.

225 commits is a lot. If possible could you squash some commits? Or maybe this should be squash committed by whoever decides to merge this.

**Comment by chaoticgd** [comment]
I found some other issues with this: a memory leak in the old SymbolGuardian implementation caused by me not being able to capture a std::unique_ptr in a std::function, and a performance issue with destroying large amounts of symbols. The new symbol guardian design should fix the first one, and the second one should be easy enough to fix.

File: pcsx2-qt/Debugger/SymbolTree/SymbolTreeWidgets.cpp
Comment (chaoticgd) [reply pcsx2-qt/Debugger/SymbolTree/SymbolTreeWidgets.cpp:368]: I'm now disabling the goto in disassembly/memory view options when it doesn't make sense to use them.

**Comment by chaoticgd** [comment]
That should be the symbol guardian issues, the performance problem, and the menu thing fixed. As for the commit history, I must admit I don't have much experience using git rebase. I could learn, although I am a bit paranoid I might screw something up, either with git or the PR itself. I'll have to do some more research into that.

**Comment by Daniel-McCarthy** [comment]
> I don't have much experience using git rebase. I could learn, although I am a bit paranoid I might screw something up, either with git or the PR itself. I'll have to do some more research into that.

I encourage you to make a branch in your local repository off of your working branch as a backup so that you have a copy of your commits before attempting a rebase. I have made a copy of the branch with your current commits here for safe keeping as well: <https://github.com/Daniel-McCarthy/pcsx2/tree/Backup-ChaoticGd-SymbolTablePR>
This way, if anything goes wrong, you can get all your commits back without a problem.

If you have questions I can help. A good place to start might be squashing together commits that might be able to be joined.
If anything goes wrong I can help you get back to where your branch was before.

**Comment by chaoticgd** [comment]
Found another regression with ScanForFunctions.

**Comment by chaoticgd** [comment]
Alright, some ideas for rebasing:
- Change the "DataInspector:" commit messages to "Debugger:".
- Remove the accidental rcheevos changes from the history.

Anything else? How aggressive do you want me to be with this? I could also go through and collapse down pairs of commits e.g. "Debugger: Rewrite symbol tree grouping algorithm" and "Debugger: Clean up the new symbol tree grouping algorithm".

I forgot to mention this before, but I also copied in some unit tests for CCC a while back with the idea that eventually people might want to work on CCC in the PCSX2 repo.

**Comment by chaoticgd** [comment]
I've been discussing how to get this merged with Dna and f0bes. Dan brought up how the updater shows the commit history for its changelog, so how merging over 200 commits might be problematic. I hence think it would be best to just do a squash merge for this one, instead of trying to rebase it.

I noticed you seem to be gearing up to do a stable release. Given the obscene size of this PR, I assume you're not going to want to merge this until that's out.

**Comment by F0bes** [comment]
> I've been discussing how to get this merged with Dan and f0bes. Dan brought up how the updater shows the commit history for its changelog, so how merging over 200 commits might be problematic. I hence think it would be best to just do a squash merge for this one, instead of trying to rebase it.
> 
> I noticed you seem to be gearing up to do a stable release. Given the obscene size of this PR, I assume you're not going to want to merge this until that's out.

I don't have merge rights here, but I don't believe we can do our regular rebase commit here, it didn't let me do so on my fork. (I assume due to the merge commits):
image

A merge commit I'm pretty sure would break the updater. I tested it on a local branch and it ends up doing stuff like the following (this is bad for us because pr commits are assumed to be bunched together):
image

Finally, a squash merge works fine locally:
image

**I think the best choice here is to do a squash commit when the maintainer merges this into master**

**Review by kamfretoz — APPROVED** [review]

**Review by weirdbeardgame — APPROVED** [review]

**Review by Daniel-McCarthy — APPROVED** [review]

**Comment by chaoticgd** [comment]
Noticed an annoying regression with the disassembler. I'll look into it.

**Comment by chaoticgd** [comment]
> Noticed an annoying regression with the disassembler. I'll look into it.

False alarm. Or rather, the wrong alarm sounded. What I noticed was that symbols sometimes weren't being substituted into the disassembly properly. I turns out this was caused by [a typo in this commit](https://github.com/PCSX2/pcsx2/commit/e9ead0d33d528417c8401224879750c725375c3b#diff-f51e885f1fdb08284246a3dffbabe3e9e335b30cf31b4fc8b08e977fc922bdaaR417). Because instructions disassembled outside of a function won't touch this code path it tricked me into thinking it was a problem with my PR.

The flurry of commits is because while investigating this issue I noticed a bunch of other problems, which actually were bugs with this PR. These are all fixed, with the exception of handling some cases where function symbols overlap (e.g. because of overlays which all get dumped into a single symbol table), although it would probably be okay to wait for a future PR to see about handling that better as DWARF support may help with that

 I also made some other miscellaneous improvements while I was at it.

**Comment by chaoticgd** [comment]
That CI failure looks like a random linker crash (bleh). Could someone with the permissions re-run it?

**Comment by stenzek** [comment]
Is it possible to squash the commits down into a smaller number? For example, one to add the new demangler, another to swap over to it, etc.

It's quite difficult to review 252 commits over hundreds of files, and it's also not rebase-mergeable (which is our preferred technique).

**Comment by stenzek** [comment]
For smaller PRs, yeah, we can just squash them on merge. But with 50k lines added, I'd prefer to split it up, to make bisecting issues easier.

**Comment by fjtrujy** [comment]
Hello @chaoticgd really nice feature!
If you need help with the rebase or squashing I can help you with that
Cheers

**Comment by chaoticgd** [comment]
> Hello @chaoticgd really nice feature! If you need help with the rebase or squashing I can help you with that Cheers

I'd like to try fixing the history myself first, I plan to resume working on it soon.

**Comment by fjtrujy** [comment]
> > Hello @chaoticgd really nice feature! If you need help with the rebase or squashing I can help you with that Cheers
> 
> I'd like to try fixing the history myself first, I plan to resume work on it soon.

Now that I see you are active again here...

One thing important to highlight when fixing the git history (I suppose you are aware). 
If you run a `rebase -i` all the "merge commits" will disappear, the more often you perform a "merge from master" the more difficult to rewrite the history will be.

**Comment by chaoticgd** [comment]
I've recreated the git commit history [here](https://github.com/chaoticgd/pcsx2/tree/data_inspector_redux). These commits have all been [ran through the CI checks](https://github.com/chaoticgd/pcsx2/pull/5).

If these new commits are satisfactory I can force push them over the data_inspector branch. [This](https://github.com/chaoticgd/pcsx2/tree/data_inspector_backup) branch will continue to point to the old history.

There is a diff between the two. The old SymbolMap class is still included in the non-rebased branch due to an error on my part. Other than that, [there are some tiny changes I decided to make while reading through the diffs](https://github.com/user-attachments/files/16753891/smalldiff.txt).

**Comment by F0bes** [comment]
Yeah, that new branch looks fine. Feel free to force push this branch. Once that's done I'll ask some people to test and we can get this merged.

**Comment by chaoticgd** [comment]
Okay, I've pushed it.

**Review by weirdbeardgame — APPROVED** [review]
Was able to read and write to ene_wrk and plyr_wrk structs in Fatal frame 1

**Review by dreamsyntax — APPROVED** [review]
Verified in GR2004;
A great improvement

**Comment by fjtrujy** [comment]
Amazing! Well done!

**Review by F0bes — APPROVED** [review]
Tested homebrew, it shows all of my globals (and reminded me of how poorly I made that thing years ago :sweat_smile:)
Tested some games. I don't have any that have the mdebug or sndata support but I did not find any regressions.
No compiler warnings.
Good work.

**Comment by lightningterror** [comment]
> No compiler warnings.

Except 2: https://github.com/PCSX2/pcsx2/actions/runs/10568682666/job/29280116405#step:12:434

**Comment by F0bes** [comment]
Oh, I checked the diff of a single commit and didn't see anything that touched that code, oops.

**Pull request merged**
